### DISCUSSION TITLE
Technologie to choose for Web API server push

### DISCUSSION THREAD

**Discussion by salokiam** [bisq-network/bisq2#278] [General]
# BISQ2 Web API needs a push technology
The REST API of Bisq2 needs also a push technology. This is an overview which choices we have.
## Data Objects to push

- private chat, Public-chat, Market chat
- Price indices as published by the price nodes?
- offer books
- trade events (offers taken, trade confirmed,...)

## Network topology

- private server, each client hosts its own server (1:1 between client and server), most probably the client and the server are on the same machine.
- the DAO may run a server for the bisq market website, this is a one-to-one relation as well.
- public hosting of server for others is not currently intended (may change)

## Available technologies

| Name                        | ease of use | scalability | support |
|-----------------------------|-------------|-------------|---------|
| WebSockets                  | medium      | medium      | high    |
| HTTP2 push                  | --          | --          | --      |
| Server-sent Events (SSE)    | medium      | medium      | medium  |
| WebTransport                | --          | --          | --      |
|                             |             |             |         |
| simple polling              | very high   | very low    | high    |
| Long polling                | very high   | medium      | high    |
| continuous polling          | medium      | medium      | medium  |
|                             |             |             |         |
| MQRabbit or other MQ        | very low    | high        | low     |
| HTTP Callbacks aka WebHooks | low         | high        | high    |

## Pros and cons

HTTP2 push is actually used only between server and browser. WebTransport is a HTTP3 technology and not yet widely available. These 2 can't be used.

WebSockets is widely used and long tme around with lots o examples. However, the novice programmer has to ramp up with the technology if he is new to the push-space.

SSE send event from the server to the client. There is a lot of support from Framework but not as much as for WebSockets.

simple polling sends requests in short intervals to see if there is data to collect. This eats up a lot of resources, but is simple to program. simple polling has a very high
latency, which is in trading never a good thing.
In long polling the client sends one request, the server waits and holds the connection until  the next data is incoming data for the client and then transfers it to the client, then ends the connection. The client the needs to submit a new request to get the next piece of data.
This only requires the client to program a loop to get continuously data.

in continuous polling the client and the server hold the connection infinitely. Messages are passed to the client as they arrive at the server. The client will need to parse the messages from the partial stream instead of parsing them after the connection terminated. Although this is not difficult, most novice programmers probably haven't one that before and this is a little hurdle.

MQRabbit is included for any Message Queue Product. These products are made for this kind of pub/sub scenarios and are most technically sophisticated. However, they are too complicated nad technical overkill for what we need.

WebHooks is a technique, where the client will register a call back URL with teh server and the server will call the client via this URL whenever there is a message. This obviously requires the client to run a webserver for the callback. Setting up a LAMP Server is standard for any level of webprogrammer but still a hurdle.

I can provide a bunch of links if someone is interested.

## Conclusion

As we don't have scalability issues in our scenario, we can go for the easiest solution. While WebSockets ha the lowest latency, it is still a technologie which the 
newbie needs to learnt first.
Long polling has a very good latency as well and is the simplest to program for the client (and for the server). No additional technology involved.
Therefore, I would recommend long polling.

If scalability becomes an issue, I would recommend looking into WebHooks, as it has only known technologies as well.

**Comment by chimp1984** [comment]
Thanks for the overview.
Have you looked into ZeroMQ as well? I guess it might be also more on the advanced side, but maybe worth considering as well.
From your overview I agree that long polling seems to be the best choice. Simplicity, ease to use and our low requirements for the 1 to  1 use case might make it the best choice.

@alkum @alvasw Do you have any feedback on that?

**Reply by salokiam** [reply]
I have used seceral MQ products before, but zeroMQ seems to be very interesting if Message Queueing is something we need. From the first looks its has an all Java version and pretty compact but powerful. But as you said, thats a more sophisticated solution. And would pull a lot of code into our product.

**Comment by alvasw** [comment]
Thank you for creating this excellent overview!

1. Is there a specific reason we're focused on web solutions?
2. Do we want to only support web technologies?
  a) What about Python/NodeJS scripts?
3. What do you think about choosing a "harder" but better solution and creating an SDK that hides all the low-level details. We provide a method where the developer can register listeners for events. In this scenario, the developer doesn't need to know anything about WebSockets, SSE, ... . (We can only officially support the most popular languages.)

**Reply by alkum** [reply]
I think this is specifically around extending the REST API, so that's why solutions are more websockets/HTTP-based.

**Reply by chimp1984** [reply]
I am not sure if a SDK would be the right attempt as it would consider the project more like a library where others build stuff/applications on top of it. I guess that is in the current state not a goal, even it would be interesting to see it in that direction or it could develop to that. The main goal of the API is to allow usage from trading scripts and allow usage in Ubrel-like setups where a headless application with some sort of lightweight usually web based client interface is a requirement. I think standard REST API should be easy to use from any programming language and the push support should add as little overhead/complexity to that model as possible. The long polling seems to fulfill that best.

**Reply by salokiam** [reply]
re 1. This part of the analysis I only looked at the psuh technologies. But the API will has a lot of functionality which is pull rather than push. So we figured out at the last dev-call that a Web REST API is the most standard and easiest way to do this. So i guess the reason we were choosing web technologies are standard and ease of use. Main focus is to enable as many developers as possible to use this API.

re 2. supporting web technologies doesnt mean you cant use Python/NodesJS. Those devs would also be able to use the (web) API.

re 3. We are still able to deliver SDKs on top on the web API for specific (most used) programming languages. And often that is being done for REST APIs to make it less work for the API users. Having the SDK as the Interface and defining the underlying technologie as implementation specific will be much more work (we need to deliver DSKs in so many languages) for us and exclude some language we cant support.

**Comment by alkum** [comment]
I agree, long polling sounds the most promising.

One thing to consider is authentication. The exposed end-points should only allow long polling and respond to authenticated calls.

**Reply by salokiam** [reply]
If we use authentication, do you also propose to use transport encryption (TLS) to protected the password? Do you mean we ness authentication in the one-to-one scenaio were client and server are on the same maschine? Or you think we should not assume that the users run both on the same maschine?

**Reply by alkum** [reply]
I think cleartext is fine for a start, as long as the REST API binds to the loopback address (e.g. only allowing access from the same machine).

**Comment by ghubstan** [comment]
@salokiam Hi, do you have a matrix acct so we can discuss this subject?  (I did a lot of study about how a rest api could be added to Bisq, about a year ago, and would like to join the conversation -- between us at first.)

**Reply by salokiam** [reply]
@ghubstan Great, I just send you a DM on matrix. Would love to talk to you.
